File: Data Structures & Algorithms/split-array-largest-sum/submission-0.py
class Solution:
    def splitArray(self, nums: List[int], k: int) -> int:
        l, r = max(nums), sum(nums)

        res = 0
        def canPartition(target):
            currSum, currSubs = 0, 1
            for num in nums:
                if currSum + num > target:
                    currSum = num
                    currSubs += 1
                else:
                    currSum += num
            return currSubs <= k

        while l <= r:
            m = (l + r) // 2
            if canPartition(m):
                res = m
                r = m - 1
            else:
                l = m + 1
        return res

        

File: Data Structures & Algorithms/subarray-product-less-than-k/submission-0.cpp
class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
        int res = 0, l = 0;
        long long product = 1;

        for (int r = 0; r < nums.size(); ++r) {
            product *= nums[r];

            while (l <= r && product >= k) {
                product /= nums[l++];
            }

            res += r - l + 1;
        }

        return res;
    }
};